lib/portfinder.js

```diff
@@ -38,9 +38,11 @@ internals.testPort = function(options, callback) {
   function onListen () {
     debugTestPort("done w/ testPort(): OK", options.host, "port", options.port);
 
-        options.server.removeListener('error', onError);
-        options.server.close();
+    options.server.removeListener('error', onError);
+    options.server.close(function () {
+      debugTestPort("done w/ testPort(): Server closed", options.host, "port", options.port);
       callback(null, options.port);
```

Comment:
just b/c the we've received a close event doesn't mean resources are fully de-allocated by the OS, it just means that node has notified the os to close, and the os has notified back to node that "go ahead and do things that don't depend on having an open file descriptor" <-- of which does not include "opening another one immediately" unless we wrap this in a try/catch or similar, and recurse - the OS may too be starved. 

in a perfect world, or with sequential consistent ordering (which is slow and x86 makes less needed) this would work.... 

I may have to verify this claim, but in general, with CSP (strongly ordered) scheduling in place, all that matters is that c comes after a and b comes after a but whether c comes after b, is often irrelevant for ordering that of the scheduler, it only cares about the eventual ordering... 

I should probably make sure my brain is not misleading me, in meantime, feel free to comment thoughts, i'll come back next week when time permits (no more than 8 days fyi)

Reply:
also, if this port really is blocked, running 2 instances of in 2 shells lets say, would yield indeterminate results as to which terminal wins, again due to os scheduling (which I suppose is to say, if you wrote a test with 2 child processes (say process x and y) trying to aquire the lock on the FD, 100 times in a row, the results, would be 50/50 as to which child process wins and gets to use port 22(in this example) - which is just another explanation for how os scheduling that is not strongly ordered (which is expensive hence the x86 optimizations to auto schedule this, without breaking eventual consistency) of the total order. 

I know its a bit confusing, this feature dropped in the late 90's IIRC... 

I will look deeper though, but wanted to add another example, i may be forgetting something that makes this easier, if so, jog my memory in the meantime :)